Assemble 1.21.30

dries-c opened the pull request:
## Introduction
This PR updates PM5 to 1.21.30

## Changes
## Backwards compatibility
See https://github.com/pmmp/BedrockProtocol/pull/270

dries-c committed:
Changes for 1.21.30

IvanCraft623 commented on src/player/Player.php:
```diff
@@ -2013,15 +2013,15 @@ public function toggleSwim(bool $swim) : bool{
 	}
 
-	public function emote(string $emoteId) : void{
+	public function emote(string $emoteId, int $emoteLengthTicks) : void{
```
This is a BC-Breaking change

IvanCraft623 commented on src/entity/Human.php:
```diff
@@ -178,10 +178,10 @@ public function jump() : void{
 	}
 
-	public function emote(string $emoteId) : void{
+	public function emote(string $emoteId, int $emoteLengthTicks) : void{
```
This is a BC-Breaking change

dries-c committed:
Remove BC breaking changes

Frosxc commented:
@IvanCraft623

dktapps commented on src/data/bedrock/block/convert/BlockStateToObjectDeserializer.php:
```diff
@@ -1593,8 +1589,7 @@ private function registerDeserializers() : void{
 			return Blocks::TNT()
-				->setUnstable($in->readBool(StateNames::EXPLODE_BIT))
-				->setWorksUnderwater($in->readBool(StateNames::ALLOW_UNDERWATER_BIT));
+				->setUnstable($in->readBool(StateNames::EXPLODE_BIT));
```
`setWorksUnderwater(false)` should be added here in case the default changes

dktapps commented on src/network/mcpe/NetworkSession.php:
```diff
@@ -784,10 +784,10 @@ function() use ($protocolVersion) : void{
-		$this->tryDisconnect(function() use ($ip, $port, $reason) : void{
-			$this->sendDataPacket(TransferPacket::create($ip, $port), true);
+		$this->tryDisconnect(function() use ($ip, $port, $reason, $reloadWorld) : void{
+			$this->sendDataPacket(TransferPacket::create($ip, $port, $reloadWorld), true);
```
What is this property for?

dries-c replied:
As far as I know it just bricks hard transfers whenever it's true. No clue what it's supposed to do

dktapps commented on src/network/mcpe/StandardEntityEventBroadcaster.php:
```diff
@@ -67,7 +67,7 @@ public function syncAttributes(array $recipients, Living $entity, array $attribu
 			$this->sendDataPacket($recipients, UpdateAttributesPacket::create(
 				$entity->getId(),
-				array_map(fn(Attribute $attr) => new NetworkAttribute($attr->getId(), $attr->getMinValue(), $attr->getMaxValue(), $attr->getValue(), $attr->getDefaultValue(), []), $attributes),
+				array_map(fn(Attribute $attr) => new UpdateAttribute($attr->getId(), $attr->getMinValue(), $attr->getMaxValue(), $attr->getValue(), $attr->getMinValue(), $attr->getMaxValue(), $attr->getDefaultValue(), []), $attributes),
```
Why are the min/max repeated?

IvanCraft623 replied:
On the protocol side there are some new fields: `float $defaultMin, float $defaultMax`, don't know what those for, maybe dries could explain

dktapps replied:
I see. Probably because the min/max can be altered by modifiers.

IvanCraft623 committed:
Normal TNT uses `setWorksUnderwater(false)` in case the default changes

IvanCraft623 committed:
Do not rebroadcast $emoteLengthTicks

Seems to be irrelevant for the client, we cannot risk rebroadcasting random values received

IvanCraft623 committed:
Merge branch 'stable' into bedrock-1.21.30

WilliamThibo2 commented:
hello, how to have pocketmine mp to the newest version?

IvanCraft623 committed:
Set `reloadWorld` property to false always

Client is not being transferred in case of `true`, maybe it's expects something else after `TransferPacket`...

IvanCraft623 committed:
Update BedrockWorldData version values
